File: eval/summarize.js
import { existsSync, readdirSync } from "fs";
import { join } from "path";

import { parseOption, print } from "./eval.util.js";
import { Logger } from "./Logger.js";


const RESULTS_DATE = parseOption("--date");
if(!RESULTS_DATE)
    throw new ReferenceError(`Missing results directory date identifier`);

const RESULTS_DIR_PATH = join(import.meta.dirname, "results", RESULTS_DATE);
if(!existsSync(RESULTS_DIR_PATH))
    throw new ReferenceError(`Results do not exist '${RESULTS_DATE}'`);

const SUMMARY_FILENAME = "__summary.json";


const summary = {};
for(let resultsFileName of readdirSync(RESULTS_DIR_PATH)) {
    if(resultsFileName === SUMMARY_FILENAME) continue;

    const resultsFilePath = join(RESULTS_DIR_PATH, resultsFileName);
    const results = (await import(resultsFilePath, { with: { type: "json" } })).default;

    summary[resultsFileName] = {
        successCases: 0,
        failureCases: 0,
        errorCases: 0,
        successRate: 0.0,
        totalSnapshotSize: 0,
        totalEstimatedTokens: 0,
        totalRTT: 0
    };
    results
        .results
        .forEach(result => {
            summary[resultsFileName].successCases += +result.success;
            summary[resultsFileName].failureCases += +!result.success;
            summary[resultsFileName].errorCases += +result.error;
            summary[resultsFileName].totalSnapshotSize += +(result.snapshotSize ?? 0);
            summary[resultsFileName].totalEstimatedTokens += +(result.tokenEstimate ?? 0);
            summary[resultsFileName].totalRTT += +(result.rtt ?? 0);
        });

    const totalCases = summary[resultsFileName].successCases
        + summary[resultsFileName].failureCases;

    const mean = (key, ignoreErrors = true) => {
        return summary[resultsFileName][key]
            / (totalCases - (ignoreErrors ? summary[resultsFileName].errorCases : 0));
    };

    summary[resultsFileName].successRate = mean("successCases", false);
    summary[resultsFileName].errorRate = mean("errorCases", false);
    summary[resultsFileName].meanSnapshotSize = mean("totalSnapshotSize");
    summary[resultsFileName].meanEstimatedTokens = mean("totalEstimatedTokens");
    summary[resultsFileName].meanRTT = mean("totalRTT");
}

new Logger(join("..", "results", RESULTS_DATE), false)
    .write(SUMMARY_FILENAME, JSON.stringify(summary, null, 2));

print(`Summary written to ${RESULTS_DATE}/${SUMMARY_FILENAME}`, true);

File: eval/_summarize.py
import json
from pathlib import Path

from eval_util import parse_option, echo
from logger import Logger


_RESULTS_ROOT = Path(__file__).parent / "results"

SUMMARY_FILENAME = "_summary.json"


def _mean(value: float, divisor: float) -> float:
    return value / divisor if divisor else 0.0


def summarize(date: str) -> dict:
    results_dir = _RESULTS_ROOT / date

    print(results_dir)

    if not results_dir.exists():
        raise ReferenceError(f"Results do not exist '{date}'")

    summary: dict = {}

    for results_file in sorted(results_dir.iterdir()):
        if results_file.name == SUMMARY_FILENAME or results_file.suffix != ".json":
            continue

        data = json.loads(results_file.read_text())
        entries = data["results"]

        s = {
            "successCases": 0,
            "failureCases": 0,
            "errorCases": 0,
            "totalSnapshotSize": 0,
            "totalEstimatedTokens": 0,
            "totalRTT": 0.0,
        }

        for r in entries:
            s["successCases"] += int(bool(r.get("success")))
            s["failureCases"] += int(not r.get("success"))
            s["errorCases"] += int(bool(r.get("error")))
            s["totalSnapshotSize"] += r.get("snapshotSize") or 0
            s["totalEstimatedTokens"] += r.get("tokenEstimate") or 0
            s["totalRTT"] += r.get("rtt") or 0

        total = s["successCases"] + s["failureCases"]
        non_error = total - s["errorCases"]

        s["successRate"] = _mean(s["successCases"], total)
        s["errorRate"] = _mean(s["errorCases"], total)
        s["meanSnapshotSize"] = _mean(s["totalSnapshotSize"], non_error)
        s["meanEstimatedTokens"] = _mean(s["totalEstimatedTokens"], non_error)
        s["meanRTT"] = _mean(s["totalRTT"], non_error)

        summary[results_file.name] = s

    summary = {
        k: v
        for k, v in sorted(
            summary.items(),
            key=lambda item: item[1]["successRate"],
            reverse=True,
        )
    }

    Logger(str(Path("..") / "results" / date), clean_dir=False).write(
        SUMMARY_FILENAME, json.dumps(summary, indent=2)
    )

    echo(f"Summary written to {date}/{SUMMARY_FILENAME}", always=True)

    return summary


if __name__ == "__main__":
    date = parse_option("--date")

    if not date:
        raise ReferenceError("Missing results date (--date <dirname_date>)")

    summarize(date)